### PR TITLE
 feature/pid-4-config-handle-for-communities

### LINKED ISSUE #96: feature/pid-1-PIDs-associated-to-metadata
## Problem description
Centers need to associate PIDs with their metadata records. These PIDs should be suitable for both human and machine. 
The class which communicates with the CNRI API must send specific Handle information to the CNRI.

Return more information to the CNRI Handle Server:
- Updated HandlePlugin class, the method getRawHandleValues. Copy code from the Clarin-DSpace. E.g.: http://hdl.handle.net/11346/113488-ZQUO?noredirect

### Dead Handle (only external handles)
- Some external monitoring is checking if the External Handle url is actual/accessible. If not it set Handle as dead.
- To the Handle CNRI is sending the metadata if the Handle is dead, e.g.: http://hdl.handle.net/11346/clarin.si-LWBQ?noredirect

### Note
- Normal Handle: http://hdl.handle.net/11234/1-4832?noredirect
- External Handle: http://hdl.handle.net/11346/113488-ZQUO?noredirect
- Dead External Handle: http://hdl.handle.net/11346/clarin.si-LWBQ?noredirect

### How to test:
This functionality couldn't be tested unless we don't have Handle.net registry testing environment.

### LINKED ISSUE #100: feature/pid-4-config-handle-for-communities
**ETA:** 38

## Problem description
- Different handle prefixes for different communities will be defined e.g. community 'Faculty of Informatics science' will have subprefix '1' that means handles in this community will look like: '<server>/handle/123456/**1**-87651.
Which prefix should have any community is defined in the configuration.
Check Clarin-DSpace documentation: [here](https://github.com/ufal/clarin-dspace/wiki/PIDs-Persistent-Identifiers#per-community-handle-prefix)
For understanding handle subprefixes check Clarin-DSpace item catalogue.
- External handles will be supported
External handles have specific subprefix - 'LRT' - EPIC handles
For External handles must be done another implementation (EPIC) of the IdentifierService. You can inspire by the Handle/DOI implementation of the IdentifierService.
Because of using external handles must be stored another value in the Handle table - add column 'URL' in the URL column will be stored external handle URL.

- Define the API which manage the external Handles which has the `@magicLindat@` string in the URL. Check the Clarin-DSpace [API](https://github.com/ufal/clarin-dspace/blob/clarin/dspace-rest/src/main/java/cz/cuni/mff/ufal/dspace/rest/MyHandleResource.java)

### PR DESCRIPTION
**Pull request by milanmajchrak**
| Phases            | MM  |  MB  | MR |   JM | MP | Total  |
|-----------------|----:|----:|-----:|-----:|-----:|-------:|
| ETA                  |  38  |    0 |     0 |      0 |      0 |        0 |
| Developing      |   4+7+4+4+1+6.5 |    0 |    0 |      0 |      8+10+6+1 |         0 |
| Review             |  0  |    0 |    0 |      0 |      0 |         0 |
| Total                |   -  |   -   |  -    |   -    |   -   |         0 |
| ETA est.             |      |       |       |         |      0 |        0 |
| ETA cust.           |   -  |   -  |   -   |   -     |      0 |        0 |
## Problem description
- Different handle prefixes for different communities will be defined e.g. community 'Faculty of Informatics science' will have subprefix '1' that means handles in this community will look like: '<server>/handle/123456/**1**-87651.
Which prefix should have any community is defined in the configuration.
Check Clarin-DSpace documentation: [here](https://github.com/ufal/clarin-dspace/wiki/PIDs-Persistent-Identifiers#per-community-handle-prefix)
For understanding handle subprefixes check Clarin-DSpace item catalogue.
- External handles will be supported
External handles have specific subprefix - 'LRT' - EPIC handles
For External handles must be done another implementation (EPIC) of the IdentifierService. You can inspire by the Handle/DOI implementation of the IdentifierService.
Because of using external handles must be stored another value in the Handle table - add column 'URL' in the URL column will be stored external handle URL.

- Define the API which manage the external Handles which has the `@magicLindat@` string in the URL. Check the Clarin-DSpace [API](https://github.com/ufal/clarin-dspace/blob/clarin/dspace-rest/src/main/java/cz/cuni/mff/ufal/dspace/rest/MyHandleResource.java)

- The [pid-1](https://github.com/dataquest-dev/DSpace/issues/96) issue is solved in this branch because here are a lot of classes which are needed to solved that issue.

### NOTE
All Actions will fail unless we won't be registered in the epic consortium: https://www.pidconsortium.net/. The tests will fail because we updated the method `createId` which is used by many methods and it is using the `*pid.service.*` properties which are not defined in the `*cfg`.
This properties must be set in the `clarin-dspace.cfg`:
- `lr.pid.service.type = `
- `lr.pid.service.url = `
- `lr.pid.service.user =`
- `lr.pid.service.pass =`

### TODO
- [x] Create classes which loads community prefix definition from configuration. Copy or modify Clarin-DSpace class and other classes which are in that package: [PIDCommunityConfiguration](https://github.com/ufal/clarin-dspace/blob/clarin/dspace-api/src/main/java/cz/cuni/mff/ufal/dspace/handle/PIDCommunityConfiguration.java)
- [x] Create PidServicePID as in the Clarin-DSpace: [PIDServiceEPICv2](https://github.com/ufal/clarin-dspace/blob/clarin/dspace-api/src/main/java/cz/cuni/mff/ufal/dspace/PIDServiceEPICv2.java)
- [x] Update handle creating process: based on the _prefix community configuration_ decide which prefix should have the item handle e.g. if in the configuration is told that community 'University' should have prefix '1' create handle with subprefix '1'. 
Add that logic here: https://github.com/dataquest-dev/DSpace/blob/dtq-dev/dspace-api/src/main/java/org/dspace/handle/HandleServiceImpl.java#L375 The HandleDAO generates the next available suffix based on the sequence `handle_seq`. Put the community suffix to that generated suffix and return it as String.
- [x] Added class PIDCommunityConfiguration for load configuration, where is info about subprefix for communities.
- [x] Used for reading configurations getArrayproperty. 
- [x] Created class PIDConfiguration.
- [x] Created test for PIDConfiguration
- [x] Created new service DspaceObjectClarinService for finding principal community. Service was added to core-services.xml.
- [x] Added DspaceObjectClarinService and HandleclarinService to ContentServiceFactory.
- [x] Created test for DspaceObjectClarinImpl
- [x] Created PIDService and AbstractPIDservice
- [x] Added class DSpaceApi
- [x] Edited HandlePlugin
- [x] Created PIDService for EPICv2
- [x] Set method for created handle based on type and community in configuration in HandleServiceImpl for local item.
- [x] Create EPIC and EPICv2 handle
- [x] Local handle - we need to change in clarin-config.cfg type from epic to local
- [x] Epic handle - HandleserviceImpl: createId calling method handle_HandleManager_createId (DSpaceApi). We probably don't have access rights. (PIDServiceEPICv2 method: sendPIDCommand) 
Test: PIDConfigTest Exception: External PID service is not working. Please contact the administrator
Because there is type epic in  clarin-config.cfg.
- [x] Create tests
- [x] Add dead and dead_since column to the Handle
- [x] Create HandleBuilder for testing
- [x] Update HandlePlugin.getRawHandleValues to associate PIDs to metadata.
- [x] Create resolveToObject, isDead, isDeadSince method in the HandleClarinService
- [x] Register in the EPIC consotrium
- [x] Register in the Handle.net registry

**External handles:**
- [x] Create API which manage the external handles
- [x] Create IT for that API
- [x] Create the new Handle Resource for External handles

## Analysis
(Write here, if there is needed describe some specific problem. Erase it, when it is not needed.)
## Problems
**1.** When the new version of the Item is creating the community suffix is not added to the handle because the method `makeIdentifierBasedOnHistory` creates the handle based on the history. This behavior will be changed in the issue https://github.com/dataquest-dev/DSpace/issues/127
**2.** The API for the External handles wasn't analyzed
**3.** The magicLindat wasn't analysed. I need to create ExternalHandleRestRepository for the Handles with magicLindat url.

### CODE REVIEW

File: dspace-api/src/main/java/org/dspace/handle/PIDConfiguration.java [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+
+        if (Objects.isNull(pidCommunityConfigurationsArray)) {
+            return;
+        }
+
+        //hashmap for creating PIDCommunityConfiguration
+        Map<String, String> map = new HashMap<String, String>();
+        pidCommunityConfigurations = new HashMap<UUID, PIDCommunityConfiguration>();
+        //exists minimally one configuration, so first community is added to map not in cycle
+        String[] keyValue = pidCommunityConfigurationsArray[0].split("=");
```
Comment (mrovnanik) [inline dspace-api/src/main/java/org/dspace/handle/PIDConfiguration.java:73]: what if there is no "=" in `pidCommunityConfigurationsArray[0]`?

**Review by mrovnanik — CHANGES_REQUESTED** [review]
Check out the comments and come with a solution

File: dspace-api/src/main/java/org/dspace/content/DspaceObjectClarinServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+        if (CollectionUtils.isEmpty(communities)) {
+            return null;
+        }
+
+        // principal community is in the first index
+        return communities.get(0);
+    }
+
+    /**
+     * The DSpaceObject has some collection where belongs - return that collection
```
[the remark this answers is not in the document]
Comment (milanmajchrak) [reply dspace-api/src/main/java/org/dspace/content/DspaceObjectClarinServiceImpl.java:52]: changed to `Return the collection where belongs current DSpaceObject`

File: dspace-api/src/main/java/org/dspace/handle/AbstractPIDService.java [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+import java.net.PasswordAuthentication;
+import java.util.Map;
+
+import org.dspace.services.ConfigurationService;
+import org.dspace.utils.DSpace;
+import org.springframework.stereotype.Component;
+
+/* Created for LINDAT/CLARIAH-CZ (UFAL) */
+/**
+ * Abstract class for PID service.
```
[the remark this answers is not in the document]
Comment (milanmajchrak) [reply dspace-api/src/main/java/org/dspace/handle/AbstractPIDService.java:20]: This class is copied from Clarin, we do not know why they did it this way, but we know this Abstract class is used only for managing EPIC handles (3th PIDs provider) and this class loads parameters from configuration for calling EPIC API.
I added this information to the comment:
`Abstract class for PID service which manages EPIC handles.
 This class loads the parameters from configuration for calling EPIC API.`

File: dspace-api/src/main/java/org/dspace/handle/HandlePlugin.java [hunk 4]
```diff
@@ -252,40 +286,58 @@ public byte[][] getRawHandleValues(byte[] theHandle, int[] indexList,
...
 
-                rawValues[i] = new byte[Encoder.calcStorageSize(hvalue)];
-                Encoder.encodeHandleValue(rawValues[i], 0, hvalue);
+            ResolvedHandle rh = null;
+            if (url.startsWith(MAGIC_BEAN)) {
+                String[] splits = url.split(MAGIC_BEAN,10);
+                url = splits[splits.length - 1];
+                // EMPTY, String title, String repository, String submitdate, String reportemail,
+                // String dataset_name, String dataset_version, String query, token is splits[8] but don't show that
+                rh = new ResolvedHandle(url, splits[1], splits[2], splits[3], splits[4], splits[5], splits[6],
```
[the remark this answers is not in the document]
Comment (milanmajchrak) [reply dspace-api/src/main/java/org/dspace/handle/HandlePlugin.java:328]: This is magicLindat code, I changed it to throw RuntimeException if the list has size less than 8.

File: dspace-api/src/main/java/org/dspace/handle/PIDConfiguration.java (hunk 1 above)
Comment (milanmajchrak) [reply dspace-api/src/main/java/org/dspace/handle/PIDConfiguration.java:73]: Added:
- if ArrayUtils.isEmpty(pidCommunityConfigurationsArray) return null;
- if cannot load property return RuntimeException.

**Review by mrovnanik — APPROVED** [review]

**Pull request merged**